Updated Amount validation conditions to not allow indefinite zeroes at the start.

sig5 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
The IOU Amount page didn't impose limits on the length of the string if zeroes are appended to the string. Hence the string can be indefinitely long if zeroes are used at the start.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/7068

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Open IOU Amount Page(Send/Receive).
2. Enter amounts starting with zeroes.
3. Test edge cases such as 0's at end and fractional values.


- [x] Verify that no errors appear in the JS console

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Try entering a string of consecutive zeroes.
2. Test if fractional values appear as expected.

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [ ] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->


https://user-images.githubusercontent.com/48352577/148817443-6f405a07-c6d9-4818-ac4d-a769734e3683.mov



#### Mobile Web

https://user-images.githubusercontent.com/48352577/148817462-0dc8efb4-4d80-45b5-b664-054e3647ac7c.mp4


#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

https://user-images.githubusercontent.com/48352577/148826431-5ec39a0a-9d9f-49b8-b5cd-4511f94c94e8.mp4

sig5 committed:
Updated regex to not allow more than one trailing zero at start.

sig5 commented:
**_I have read the CLA Document and I hereby sign the CLA_**

parasharrajat commented:
@sig5  Please add videos for the rest of the platforms.

sig5 commented:
@parasharrajat I don't have access to a Macbook at the moment. But since this is just a regex change, this should work just fine. Can we do without the Mac/iOS videos?

parasharrajat commented:
That's fine.

sig5 commented:
@parasharrajat Thanks :D ! Added for Android.

parasharrajat reviewed (approved):
~LGTM.~. Found an issue below.

parasharrajat reviewed (changes requested):
Found an issue 
Enter this value  => 2000222

Now place the cursor after 2 and try to delete the preceding 2. We can't delete it.

sig5 commented:
Ah, I missed that! As a user, he/she might want to change 5001 to 1001, So we should allow trailing zeroes at the beginning. 
Now what we can do is:
I can revert the regex change and modify the validation condition.
The issue comes due to the fact that the last length checking condition parses the string as a number and compares the length.
https://github.com/Expensify/App/blob/6803d0091ce4bcf68ca847818ca4a76ca3778ebe/src/pages/iou/steps/IOUAmountPage.js#L141-L144

We can instead parse the string directly after removing the Commas.
`this.stripCommaFromAmount(amount).length <= CONST.IOU.AMOUNT_MAX_LENGTH;`

And this should take care of all the values.

cc: @parasharrajat

parasharrajat commented:
Sound like a plan...

sig5 committed:
revert regex change

sig5 committed:
Updated validation condition on amount string

parasharrajat commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -140,7 +140,8 @@ class IOUAmountPage extends React.Component {
     validateAmount(amount) {
         const decimalNumberRegex = new RegExp(/^\d+(,\d+)*(\.\d{0,2})?$/, 'i');
-        return amount === '' || (decimalNumberRegex.test(amount) && (parseFloat((amount * 100).toFixed(2)).toString().length <= CONST.IOU.AMOUNT_MAX_LENGTH));
+        const lengthBeforeDecimal = this.stripCommaFromAmount(amount.split('.')[0]).length;
+        return amount === '' || (decimalNumberRegex.test(amount) && decimalNumberRegex.test(amount) && lengthBeforeDecimal <= (CONST.IOU.AMOUNT_MAX_LENGTH);
```
Why do need to check the decimal part only? It seems to be that we just need to strip `,` and `.` from the number. But in reality, the value will never have the `,`.

sig5 replied:
Because the regex enforces the number of values after the decimal to `<= 2.` If the condition is imposed on the whole string instead, user can have `500000` and `5000.0` in his valid range but not `5000.24` if  ` IOU.AMOUNT_MAX_LENGTH =6`. Applying check on only part before decimal will ensure a uniform range for the user. 

Also commas were kept because they were expected to be handled in the existing code. I will happily remove them if needed (We might be inserting commas if we are copy-pasting the amount)!

parasharrajat replied:
Yeah, commas are added. Missed that upper line.  On Backend, we check the whole number via converting the floats to integer  `(amount * 100)` is used. 
If we try the logic from the previous code. 
1. 5000.24 => 500024
2. 5000.0  => 500000

I will try to stick to the old code. and we can just prevent zeros.

Let me know if I am missing something.

sig5 replied:
The logic that was previously provided works well for a processed string (as we get in backend ) since we can be happy without caring about zeroes in the beginning. In this case, we will need to add an additional condition that essentially checks if (length of stripped portion as a parsedFloat + length of trailing zeroes ) doesn't exceed the limit.

**Previous Code:**
00500.23  => disregards the initial two zeroes, applies the older condition, and checks from the amount string if the number of zeroes, in the beginning, invalidates the amount string.

**Newer Logic**
00500.23 => takes care of the length of the string directly. The two conditions; (1) fractional part to 2 decimal places (2) Length of string before decimal should be bounded by Amount_Length.

In my humble opinion, this condition will ensure a lesser overhead, cleaner, and more generally applicable too! Do let me know your thoughts on it :D .

parasharrajat replied:
> disregards the initial two zeroes, applies the older condition, and checks from the amount string if the number of zeroes in the beginning, invalidates the amount string.

Hmm. Still, I am not sure of what you explained above but I got the idea.

sig5 replied:
Since we essentially cast strings as integers, we lose the zeroes in the beginning of the number that causes the referenced issue ( Since parseFloat("00000005")=5 ).
I am suggesting that if we directly use the string for validation, we won't be losing the zeroes in the beginning.

parasharrajat commented on src/CONST.js:
```diff
@@ -474,7 +474,7 @@ const CONST = {
             REQUEST: 'request',
         },
-        AMOUNT_MAX_LENGTH: 10,
+        AMOUNT_MAX_LENGTH: 7,
```
```suggestion
        AMOUNT_MAX_DECIMAL_LENGTH: 7,
```

sig5 replied:
We can probably do something like  ` AMOUNT_INTEGER_PART_LIMIT`

Julesssss replied:
Can you revert the value to 10 please

sig5 committed:
Updated Suggested changes

sig5 committed:
Update src/pages/iou/steps/IOUAmountPage.js

Co-authored-by: Rajat Parashar <parasharrajat@users.noreply.github.com>

sig5 committed:
updated constant name

parasharrajat reviewed (dismissed):
LGTM. 

cc: @Julesssss 

:ribbon: :eyes: :ribbon:  C+ reviewed

Julesssss commented:
Hi @sig5. please merge main to resolve the failing test.

sig5 committed:
Merge branch 'Expensify:main' into fix-issue#7068

sig5 commented:
Done. Thanks!

Julesssss commented on src/CONST.js:
```diff
@@ -476,7 +476,7 @@ const CONST = {
             REQUEST: 'request',
         },
-        AMOUNT_MAX_LENGTH: 10,
+        AMOUNT_INTEGER_PART_MAX_LEN: 7,
```
I commented on a out of date thread, so leaving it here too: 

Can you reset the value to 10 please, we specifically want to support very large IOU amounts.

sig5 replied:
We are still supporting the same range of values. The purpose of this constant has been changed now as per the validation conditions. This represents the integer part length of the amount now. So we have  (10 -3) = 7 as the value. We were supporting 10 digit strings before and still are! Now we have `AMOUNT_INTEGER_PART_MAX_LEN=7` digits before the decimal point And `2` after the decimal point. So we still support the max length of string = 10.

Julesssss replied:
Ah I didn't read it properly... That makes sense, thanks for explaining.

Julesssss reviewed (commented):
One thing I noticed is that input has a noticeable render delay with this change.

**With Changes**
https://user-images.githubusercontent.com/10736861/148991032-b714ffc9-6a22-4194-a384-1847e3349be7.mov


**Current `main`**
https://user-images.githubusercontent.com/10736861/148991646-23381dc0-b39c-43dc-8c75-522813b37b6a.mov

Julesssss reviewed (changes requested):
Holding the PR for now, to see if we can resolve the above.

sig5 commented:
@Julesssss You probably added the same videos for both cases.

Julesssss commented:
@sig5 sorry, here is the screencast from `main`

https://user-images.githubusercontent.com/10736861/148995500-5253fb3b-c812-4afa-9e07-db5a39b18ab6.mov

sig5 commented:
https://user-images.githubusercontent.com/48352577/149001351-d2d74615-fbfd-4832-b642-ac7a1c7b0b05.mp4

First Tab: New code
Second Tab : Main Code

@Julesssss Imho I have been looking at it on my setup and the rendering speeds look quite comparable. Moreover, the complexity overheads for the checks haven't really increased that much.
@parasharrajat Did you see any noticeable difference too?

sig5 commented:
Hey @Julesssss I have been looking into the code. It seems that the function `stripCommaFromAmount` is taking too much time while running. This might be the cause of slow rendering in your setup. 
Also replacing the regex based `str.replace`:

```
    stripCommaFromAmount(amount) {
        return amount.replace(/,/g, '');
    }
```
with a standard character substitution: 
```
    stripCommaFromAmount(amount) {
        return amount.replace('.', '');
//should take half the time.
    }
```
brings down the runtime to roughly half or less. I verified it here : [jsfiddle](https://jsfiddle.net/sakarq/a15tb9c8/2/).If we do this,we should be able to achieve better performance than both of the versions, the main and the new one.

sig5 committed:
revert to initial state.

sig5 committed:
added condition to handle trailing zroes.

sig5 committed:
Merge branch 'fix-issue#7068' of github.com:sig5/App into fix-issue#7068

sig5 committed:
updated constant name

sig5 committed:
Update IOUAmountPage.js

revert newline change

sig5 committed:
Updated constant name

sig5 committed:
Merge branch 'fix-issue#7068' of github.com:sig5/App into fix-issue#7068

sig5 committed:
Updated return type

sig5 commented:
I suspect `stripCommafromAmount` might be the culprit here, I have reverted to initial logic and added a check to verify the number of trailing zeroes and handle accordingly. I logged the runtimes of the function in my system and found that it is faster than the former solution.
 @Julesssss Could you please test if the rendering delay still persists in your system? Thanks!

sig5 committed:
fixed linting issue

sig5 committed:
added comments

sig5 committed:
Edit comments.

sig5 committed:
Edit comments.

Julesssss reviewed (commented):
Yep, that has fixed the lag on my size. Thanks.

It looks like there's a trailing space lint failure to fix, but other than that the PR looks good.

parasharrajat commented:
Seems, PR is changed completely. I will do the review shortly...

sig5 committed:
Fixed linting

parasharrajat commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -132,6 +133,25 @@ class IOUAmountPage extends React.Component {
+     */
+    amountLength(amount) {
+        const trailingZeroes = amount.match(/^0+/);
+        const trailingZeroesLength = trailingZeroes ? _.first(trailingZeroes).length : 0;
```
```suggestion
        const trailingZeroesLength = lodashGet(trailingZeroes, '[0].length', 0);
```

parasharrajat commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -132,6 +133,25 @@ class IOUAmountPage extends React.Component {
+     * @returns {Number}
+     */
+    amountLength(amount) {
+        const trailingZeroes = amount.match(/^0+/);
```
Should this be `leadingZeroes` ? if so, please replace all the trailing to leading.

sig5 replied:
Yeah my bad!

parasharrajat commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -132,6 +133,25 @@ class IOUAmountPage extends React.Component {
+        const absAmount = parseFloat((amount * 100).toFixed(2)).toString();
+
+        /*
+        We return sum of trailing zeroes length and absAmount length( which has value of the all digits after the decimal).
+        In case of zeroes, we add the trailing zeroes length to 2, which
+        represent places reserved for digits after decimal.
+        */
+        return trailingZeroesLength + (absAmount === '0' ? 2 : absAmount.length);
```
```suggestion
        Return the sum of leading zeroes length and absolute amount length(including fraction digits). 
        When the absolute amount is 0, add 2 to the leading zeroes length to represent fraction digits.
        */
        return trailingZeroesLength +  absAmount === '0' ? 2 : absAmount.length;
```

sig5 replied:
I addressed the comments, but am doing away with parentheses since === and ?: has lower precedence than +, it will break the code to put w/o it either at beginning or end.

parasharrajat replied:
Oh sorry, I didn't know It was left accidentally on the suggestion. I meant to request only comment update

parasharrajat commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -132,6 +133,25 @@ class IOUAmountPage extends React.Component {
+     * @param {String} amount
+     * @returns {Number}
+     */
+    amountLength(amount) {
```
```suggestion
    calculateAmountLength(amount) {
```

parasharrajat commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -132,6 +133,25 @@ class IOUAmountPage extends React.Component {
     }
 
+    /**
+     * Return length of Amount string without commas.
+     *
```
```suggestion
```

parasharrajat reviewed (changes requested):
Good work. Loved the math.

sig5 committed:
updated comments and var_names

sig5 commented:
Thanks, @parasharrajat  :D! I addressed the suggestions.

parasharrajat reviewed (commented):
@sig5 Could you please do a final checkout on performance? just to make sure lodashGet is not slow... :smile:

sig5 commented:
Hey @parasharrajat , the runtimes of both the methods seem pretty much equal on my system!

parasharrajat commented:
https://measurethat.net/Benchmarks/Show/16725/0/lodashget-vs-property-df

Native code is 1.5 times faster than lodashGet. But I am fine as far as it does not affect the UX. I leave this to @Julesssss to check. For me, both are the same and I don't see a recognizable delay.

parasharrajat reviewed (approved):
LGTM.

cc: @Julesssss

Julesssss commented:
Hi @sig5, with the latest changes I'm once again able to add multiple zeros prior to the decimal place. 

![Screenshot 2022-01-17 at 14 42 36](https://user-images.githubusercontent.com/10736861/149790304-54a12fe9-0bcb-467c-aa69-2486338dc2bf.png)

Julesssss reviewed (changes requested):
Requesting further changes.

sig5 commented:
Hi @Julesssss  this is expected since we decided to only limit the number of zeroes rather than limiting the zeroes to one since it will pose issues while editing the number. (Eg: 50001->00001-> 10001). 
The initial issue was the addition of unlimited zeroes in the number which has been resolved, So you cannot add more zeroes indefinitely but upto 8 0's .
https://github.com/Expensify/App/pull/7104#pullrequestreview-848281020
> Ah, I missed that! As a user, he/she might want to change 5001 to 1001, So we should allow trailing zeroes at the beginning.

Julesssss commented:
Ah, that makes sense. Thanks for pointing out the comment.

OSBotify commented:
@sig5, Great job getting your first Expensify/App pull request over the finish line! :tada:

I know there's a lot of information in our [contributing guidelines](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md), so here are some points to take note of :memo::

1. Now that your first PR has been merged, you can be hired for another issue. Once you've completed a few issues, you may be eligible to work on more than one job at a time. 
2. Once your PR is deployed to our staging servers, it will undergo quality assurance (QA) testing. If we find that it doesn't work as expected or causes a regression, you'll be responsible for fixing it. Typically, we would revert this PR and give you another chance to create a similar PR without causing a regression.
3. Once your PR is deployed to _production_, we start a 7-day timer :alarm_clock:. After it has been on production for 7 days without causing any regressions, then we pay out the Upwork job. :moneybag:

So it might take a while before you're paid for your work, but we typically post multiple new jobs every day, so there's plenty of opportunity. I hope you've had a positive experience contributing to this repo! :blush:

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1714969452) to staging by @Julesssss in version: 1.1.30-4 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mvtglobally commented:
@marcaaron @chiragsalian @mallenexpensify When we have a single zero upfront, it is not being stripped. Can you please confirm its expected? 
![](https://utest-dl.s3.amazonaws.com/testRunStepMinAttachmentsResult/17071801/Screen_Shot_2022-01-18_at_8.51.40_PM.png?AWSAccessKeyId=AKIAJ2UIWMJ2OMC3UCQQ&Expires=1674174750&Signature=%2Beq97xxi8oFVMQ%2F8%2Btvkz2M87BQ%3D)
![](https://utest-dl.s3.amazonaws.com/testRunStepMinAttachmentsResult/17071801/Screen_Shot_2022-01-18_at_8.51.19_PM.png?AWSAccessKeyId=AKIAJ2UIWMJ2OMC3UCQQ&Expires=1674174750&Signature=nI6HN9IPeY%2FkaZyNbjRLcjUCL6s%3D)

sig5 commented:
Hi @mvtglobally I feel like I should mention here that this was discussed here:https://github.com/Expensify/App/pull/7104#pullrequestreview-848281020,
We concluded that, in case the user just missed adding a number before the zeroes, he might want to go back and edit it with the zeroes .

 Example:
- User wants to type `60001`
- User types `50001` by mistake
- User clears the `5`, and `0001` gets stripped to `1`
- So it can pose issues while editing the number.

chiragsalian commented:
Hi @mvtglobally, yes i believe this is fine. From my understanding the code here does not allow indefinite 0's. i.e., you can't enter more than 8 0's to destroy the UI.  You can enter `00012345` and its totally fine. Its just that you cannot enter like 30 0's to break the UI (which currently happens on production). Can you retest and let us know if its a pass or if you have any more questions?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1726158569) to production by @chiragsalian in version: 1.1.31-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
